tools/bigquery-zos-mainframe-connector/src/main/scala/com/google/cloud/gszutil/CopyBook.scala

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2019 Google LLC All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.gszutil
+
+import com.google.cloud.gszutil.Decoding.{CopyBookField, CopyBookLine, Decoder, parseCopyBookLine}
+import org.apache.orc.TypeDescription
+import org.apache.orc.TypeDescription.Category
+
+import scala.collection.mutable.ArrayBuffer
+
+
+case class CopyBook(raw: String) {
+  final val Fields: Seq[CopyBookLine] = raw.lines.flatMap(parseCopyBookLine).toSeq
+
+  final val FieldNames: Seq[String] =
+    Fields.flatMap{
+      case CopyBookField(name, _) =>
+        Option(name.replaceAllLiterally("-","_"))
+      case _ =>
+        None
+    }
+
+  final val decoders: Array[Decoder] = {
+    val buf = ArrayBuffer.empty[Decoder]
+    Fields.foreach{
+      case CopyBookField(_, decoder) =>
+        buf.append(decoder)
+      case _ =>
+    }
+    buf.toArray
+  }
+
+  final val ORCSchema: TypeDescription = {
+    val schema = new TypeDescription(Category.STRUCT)
+    FieldNames
+      .zip(decoders)
+      .foreach{f =>
+        schema.addField(f._1, f._2.typeDescription)
+      }
+    schema
+  }
+
+  final val LRECL: Int = decoders.foldLeft(0){_ + _.size}
```

Comment:
I don't like to sum that way because if the collection is large it produces a bunch of intermediate garbage. That doesn't really matter here since in practice there are less than 100 fields in a table.

Reply:
Tell me what numbers you get from running this test:
```
    def test1(n: Int) = {
      System.out.println(s"testing fold...")
      (0 until n).foreach{i =>
        val sum = (0 until Int.MaxValue).foldLeft(0){_+_*0}
        System.out.println(s"$i $sum")
      }
    }

    def test2(n: Int) = {
      System.out.println(s"testing map...")
      (0 until n).foreach{i =>
        val sum = (0 until Int.MaxValue).map(_ * 0).sum
        System.out.println(s"$i $sum")
      }
    }

    val t0a = System.currentTimeMillis()
    test1(2)
    val t1a = System.currentTimeMillis()
    val dta = t1a-t0a
    System.out.println(s"fold: $dta")

    val t0 = System.currentTimeMillis()
    test2(2)
    val t1 = System.currentTimeMillis()
    val dt = t1-t0
    System.out.println(s"map: $dt")
```

Reply:
Good point. Thanks for the example!